File: src/main/java/org/genyris/core/Atom.java
package org.genyris.core;

import org.genyris.exception.AccessException;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.Environment;
import org.genyris.interp.ExpressionEnvironment;

public abstract class Atom extends ExpWithEmbeddedClasses implements Comparable {

	public abstract void acceptVisitor(Visitor guest) throws GenyrisException;
	public abstract Exp eval(Environment env) throws GenyrisException;

	public abstract String toString();

	public abstract Symbol getBuiltinClassSymbol(Internable table);

	public boolean isPair() {
		return false;
	}
	public int length(Symbol NIL) throws AccessException {
		throw new AccessException("attempt to take length of atom: "
				+ this.toString());
	}

	public Exp nth(int number, Symbol NIL) throws AccessException {
		throw new AccessException("attempt to take nth " + number + " of atom: "
				+ this.toString());
	}

	public Exp car() throws AccessException {
		throw new AccessException("attempt to take left of non-pair: "
				+ this.toString());
	}

	public Exp cdr() throws AccessException {
		throw new AccessException("attempt to take right of non-pair: "
				+ this.toString());
	}

	public Exp setCar(Exp exp) throws AccessException {
		throw new AccessException("attempt to set left of non-pair");
	}

	public Exp setCdr(Exp exp) throws AccessException {
		throw new AccessException("attempt to set left of non-pair");
	}

	public Environment  makeEnvironment(Environment parent) throws GenyrisException {
		return new ExpressionEnvironment(parent, this);
	}

	public int compareTo(Object other) {
		if(this.getClass() != other.getClass()) {
			return -1;
		} else {
			return this.toString().compareTo(((Atom) other).toString());
		}
	}
}

File: src/org/genyris/core/Atom.kt
package org.genyris.core

import org.genyris.exception.AccessException
import org.genyris.exception.GenyrisException
import org.genyris.interp.Environment
import org.genyris.interp.ExpressionEnvironment

abstract class Atom : ExpWithEmbeddedClasses(), Comparable<Any?> {
    @kotlin.Throws(GenyrisException::class)
    abstract override fun acceptVisitor(guest: Visitor?)

    @kotlin.Throws(GenyrisException::class)
    abstract override fun eval(env: Environment?): Exp?

    abstract override fun toString(): String

    abstract override fun getBuiltinClassSymbol(table: Internable?): Symbol?

    override fun isPair(): Boolean {
        return false
    }

    @kotlin.Throws(AccessException::class)
    override fun length(NIL: Symbol?): Int {
        throw AccessException(
            "attempt to take length of atom: "
                    + this.toString()
        )
    }

    @kotlin.Throws(AccessException::class)
    override fun nth(number: Int, NIL: Symbol?): Exp? {
        throw AccessException(
            ("attempt to take nth " + number + " of atom: "
                    + this.toString())
        )
    }

    @kotlin.Throws(AccessException::class)
    override fun car(): Exp? {
        throw AccessException(
            "attempt to take left of non-pair: "
                    + this.toString()
        )
    }

    @kotlin.Throws(AccessException::class)
    override fun cdr(): Exp? {
        throw AccessException(
            "attempt to take right of non-pair: "
                    + this.toString()
        )
    }

    @kotlin.Throws(AccessException::class)
    override fun setCar(exp: Exp?): Exp? {
        throw AccessException("attempt to set left of non-pair")
    }

    @kotlin.Throws(AccessException::class)
    override fun setCdr(exp: Exp?): Exp? {
        throw AccessException("attempt to set left of non-pair")
    }

    @kotlin.Throws(GenyrisException::class)
    override fun makeEnvironment(parent: Environment?): Environment? {
        return ExpressionEnvironment(parent, this)
    }

    override fun compareTo(other: Any): Int {
        if (this.getClass() != other.getClass()) {
            return -1
        } else {
            return this.toString().compareTo((other as Atom).toString())
        }
    }
}
